### PR TITLE
net: buf: linearize: Get rid of useless memset()

### PR DESCRIPTION
**Pull request by pfalcon**
net_buf_linearize() used to clear the contents of output buffer,
just to fill it with data as the next step. The only effect that
would have is if less data was written to the output buffer. But
it's not reliable for a caller to rely on net_buf_linearize() for
that, instead callers should take care to handle any conditions
like that themselves. For example, a caller which wants to process
the data as zero-terminated string, must reserve a byte for it
in the output buffer explicitly (and set it to zero).

The only in-tree user which relied on clearing output buffer was
wncm14a2a.c. But either had buffer sizes calculated very precisely
to always accommodate extra trailing zero byte (without providing
code comments about this), or arguably could suffer from buffer
overruns (at least if data received from a modem was invalid and
filled up all destination buffer, leaving no space for trailing
zero).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

File: drivers/modem/wncm14a2a.c [hunk 1]
```diff
@@ -960,10 +981,12 @@ static void on_cmd_sockdataind(struct net_buf **buf, u16_t len)
 
 static void on_cmd_socknotifyev(struct net_buf **buf, u16_t len)
 {
-	char value[40];
+	char value[41];
```
Comment (jukkar) [inline drivers/modem/wncm14a2a.c]: Why changing the length of the buffer here, was there an error in the original code?
There was a similar case in on_cmd_sockwrite(), on_cmd_sockdial() and on_cmd_sockexterror() where you did not increase the size of the value[] buffer.

**Comment by pfalcon** [comment]
As suggested by @lpereira in https://github.com/zephyrproject-rtos/zephyr/pull/12230#discussion_r245373047 .

File: drivers/modem/wncm14a2a.c (hunk 1 above)
Comment (pfalcon) [reply drivers/modem/wncm14a2a.c]: > Why changing the length of the buffer here, was there an error in the original code?

I have suspicions regarding the original code. I didn't change other buffers, because there're clearly destined for numeric data. Even if there was a thinko in the original code, and we'll go from 8 decimal digits to 7, it's a difference between (almost) 10 millions and (almost) 100 millions, and hopefully length/id's encoded there won't hit those values (which would be truncated). And going from char[8] to char[9] could increase stack usage considerably (e.g. 2x) for those functions.

Here however we have the whole command line, and I don't want to risk it, and go conservative with allowing (still) 40 actual chars.

All in all, the original author of this code, @mike-scott, is in reviewers, and certainly will correct me if anything.

Comment (mike-scott) [reply drivers/modem/wncm14a2a.c]: 40 size is ok here.  Max incoming value would be:
"LTIME",yy/mm/dd,hh:mm:ss±zz
For a total of 31 bytes (including null)

File: drivers/modem/wncm14a2a.c [hunk 2]
```diff
@@ -562,29 +562,40 @@ static void on_cmd_atcmdecho_nosock(struct net_buf **buf, u16_t len)
...
 
 static void on_cmd_atcmdinfo_revision(struct net_buf **buf, u16_t len)
 {
-	net_buf_linearize(ictx.mdm_revision, sizeof(ictx.mdm_revision),
-			  *buf, 0, len);
+	size_t out_len;
+
+	out_len = net_buf_linearize(ictx.mdm_revision,
+				    sizeof(ictx.mdm_revision) - 1,
+				    *buf, 0, len);
```
Comment (mike-scott) [inline drivers/modem/wncm14a2a.c]: Missing null termination here

File: drivers/modem/wncm14a2a.c [hunk 3]
```diff
@@ -562,29 +562,40 @@ static void on_cmd_atcmdecho_nosock(struct net_buf **buf, u16_t len)
...
 
 static void on_cmd_atcmdinfo_model(struct net_buf **buf, u16_t len)
 {
-	net_buf_linearize(ictx.mdm_model, sizeof(ictx.mdm_model),
-			  *buf, 0, len);
+	size_t out_len;
+
+	out_len = net_buf_linearize(ictx.mdm_model,
+				    sizeof(ictx.mdm_model) - 1,
+				    *buf, 0, len);
```
Comment (mike-scott) [inline drivers/modem/wncm14a2a.c]: Missing null termination here

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -531,7 +531,11 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 		recv_len = max_len;
 	}
 
-	net_frag_linearize(buf, recv_len, pkt, header_len, recv_len);
+	/* Length passed as arguments are all based on packet data size
+	 * and output buffer size, so return value is invariantly == recv_len,
+	 * and we just ignore it.
+	 */
+	(void)net_frag_linearize(buf, recv_len, pkt, header_len, recv_len);
```
Comment (mike-scott) [inline subsys/net/lib/sockets/sockets.c]: Are you sure you don't want to handle -ENOMEM errors here?

**Review by mike-scott — CHANGES_REQUESTED** [review]

File: subsys/net/lib/sockets/sockets.c (hunk 4 above)
Comment (mike-scott) [reply subsys/net/lib/sockets/sockets.c]: Hmm..  looking at code:
- You won't get the first -ENOMEM because you're passing recv_len for both length params
- You probably won't get second -ENOMEM because recv_len is set via net_pkt_appdatalen() which is usually set via net_buf_frags_len().  This function iterates through the buf frags in much the same was as linearize.

TL;DR: Pretty small chance of an error here.

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > Are you sure you don't want to handle -ENOMEM errors here?

There won't be any -ENOMEM's, sorry if I didn't cc: you on https://github.com/zephyrproject-rtos/zephyr/pull/12230 .

File: drivers/modem/wncm14a2a.c (hunk 2 above)
Comment (pfalcon) [reply drivers/modem/wncm14a2a.c]: Weird, I only may imagine that pressed "move selection" key instead of "copy selection" in my editor, because I remember carefully adding (well, copy-pasting) it to each case there. Will fix tomorrow.

Comment (pfalcon) [reply drivers/modem/wncm14a2a.c]: fixed

File: drivers/modem/wncm14a2a.c (hunk 1 above)
Comment (pfalcon) [reply drivers/modem/wncm14a2a.c]: > 40 size is ok here.

Reverted back to 40.

File: drivers/modem/wncm14a2a.c (hunk 3 above)
Comment (jukkar) [reply drivers/modem/wncm14a2a.c]: Shouldn't we have `ictx.mdm_model[out_len] = 0;` here?

File: drivers/modem/wncm14a2a.c (hunk 2 above)
Comment (jukkar) [reply drivers/modem/wncm14a2a.c]: Shouldn't we have ictx.mdm_revision[out_len] = 0; here?

File: drivers/modem/wncm14a2a.c (hunk 3 above)
Comment (pfalcon) [reply drivers/modem/wncm14a2a.c]: > Shouldn't we have ictx.mdm_model[out_len] = 0; here?

Sorry, I missed to push fixes, done now.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by mike-scott — APPROVED** [review]
